### PR TITLE
feat(discovery): decouple npmjs follower & staging

### PR DESCRIPTION
**Pull request by RomainMuller**
Separating the npmjs follower from staging allows for more efficient
tracking against the npmjs catalog, as the downloading and staging of
tarballs is parallelized more efficiently in multiple processes instead
of a single node process' libuv thread pool.

The two functions are linked together using an SQS queue, which serves
as a batching mechanism, and the staging function has a
dead-letter-queue so that "poison pill" messages are effectively
side-lined for manual review.

The new npmjs follower maintains state about which versions it has
already processed in the marker file (which auto-converts from the old
to the new format), so that it avoids re-notifying about package
versions that have not changed since they were last staged. It notifies
about ALL new package versions, not just the latest one, as was
previously the case, so that the Construct Hub will injest ALL
compatible and relevant versions of packages as part of a back-fill.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: src/backend/discovery/npm-catalog-follower.lambda.ts [hunk 1]
```diff
@@ -58,34 +53,39 @@ export async function handler(event: ScheduledEvent, context: Context) {
...
+  return new Promise<void>((ok, ko) => {
     db.changesReader.get(config)
       .on('batch', metricScope((metrics) => async (batch: readonly Change[]) => {
-        metrics.setNamespace('ConstructHub/Discovery');
+        metrics.setNamespace(METRIC_NAMESPACE);
         const startTime = Date.now();
         try {
           console.log(`Received a batch of ${batch.length} element(s)`);
-          metrics.putMetric('BatchSize', batch.length, Unit.Count);
+          metrics.putMetric(METRIC_NAME_BATCH_SIZE, batch.length, Unit.Count);
```
Comment (NetaNir) [inline src/backend/discovery/npm-catalog-follower.lambda.ts:63]: @RomainMuller its super hard to follow the changes in this PR cause it mixed refactoring and splitting the functionalities

**Review by NetaNir — COMMENTED** [review]
There is a lot in this PR so before we do a full review I want us to flush out the details of this change. As you mentioned before, the discovery function two different mode of operation - live change following and backfilling, compels it to have some weird code path which are not easy to reason with. I also think that the fact we can't scale the discovery function in the current design warrant further discussion on the design (it could be that there isn't a better way). 

I have some questions on the versioning use case, I want to make sure I understand all the places this is required in. 

Lets brainstorm this a bit before the review

File: src/backend/discovery/npm-catalog-follower.lambda.ts [hunk 2]
```diff
@@ -123,21 +123,51 @@ export async function handler(event: ScheduledEvent, context: Context) {
    *
    * @returns the value of the last transaction marker.
    */
-  async function loadLastTransactionMarker(defaultValue: number): Promise<number> {
+  async function loadLastTransactionMarker(defaultValue: number): Promise<Marker> {
+    const resetObject = { Bucket: stagingBucket, Key: RESET_BEACON_KEY };
+    const reset = await aws.s3().headObject(resetObject)
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply src/backend/discovery/npm-catalog-follower.lambda.ts:128]: Here, we don't care about the object's content, we only care about whether it exists or not. The `RESET_BEACON` can be used to reset the CouchDB follower without being annoyed by a currently running follower (which would re-create the object according to current state all the time).

File: src/backend/discovery/npm-catalog-follower.lambda.ts [hunk 3]
```diff
@@ -123,21 +123,51 @@ export async function handler(event: ScheduledEvent, context: Context) {
...
+            return value;
+          }
+          return Object.entries(value).reduce(
+            (map, [pkg, time]) => map.set(pkg, new Date(time as number)),
+            new Map<string, Date>(),
+          );
+        },
+      ) as number | Marker;
+      if (typeof result === 'number') {
+        console.log('Migrating transaction marker to new format...');
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply src/backend/discovery/npm-catalog-follower.lambda.ts:160]: This is a muscle that is worthy of exercising, though. I'm all for dropping the auto-conversion later (once this as rolled out to "production" on our pipelines), but I would like to always keep our deployments 100% human-interaction-free.

File: src/backend/discovery/index.ts [hunk 4]
```diff
@@ -63,40 +75,187 @@ export class Discovery extends Construct {
...
+    });
+
     // Note: the handler is designed to stop processing more batches about 2 minutes ahead of the timeout.
     const timeout = Duration.minutes(15);
-    const lambda = new Handler(this, 'Default', {
-      description: 'Periodically query npm.js index for new construct libraries',
+    this.npmCatalogFollower = new NpmCatalogFollower(this, 'Default', {
+      description: '[Discovery/NpmCatalogFollower] Periodically query npm.js index for new construct libraries',
       memorySize: 10_240,
       reservedConcurrentExecutions: 1, // Only one execution (avoids race conditions on the S3 marker object)
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply src/backend/discovery/index.ts:89]: The fact the catalog follower is set to 1 is to avoid a race condition on S3 with is rather difficult to work with. To be honest, I don't know that this particular work can or needs to be scaled up... Currently, a full back-fill from the "origin of CDK" takes about 3 hours, which is relatively fast considered the amount of packages being traversed...

[the remark this answers is not in the document]
Comment (eladb) [reply src/backend/discovery/index.ts:89]: I agree with @NetaNir. We need to be able to scale this by design. 3 hours sounds like **a lot**.

We want to be able to easily update our processing pipeline to produce different outputs, especially given we are moving a lot of rendering to the backend. I'd expect almost every few versions (especially until we GA) to have some changes to that pipeline that will require backfill.

Ideally We should be able to increase concurrency so we can back fill within minutes or less.

Sounds like a design constraint that blocks us from that is not healthy.

[the remark this answers is not in the document]
Comment (RomainMuller) [reply src/backend/discovery/index.ts:89]: There have been 12,840,659 package changes between the beginning of CDK (seq 1,800,000) and now (seq 14,640,659). If we can process those changes in 3 hours, give or take, this puts us at an average of 1.18ms per change. How is this too slow? What would be an acceptable threshold?

How do you propose we parallelise following a time-stream that tracks a sparse sequence of events (the above calculation is a bit twisted, as CouchDB has collapsed many of these changes and will only present certain sequence numbers, not all of them) -- other than doing precisely what this PR introduces:

1. A lambda function whose job is ONLY to traverse the change stream as quickly as possible, extracting the relevant data & notifying
2. A second lambda function (of which up to 1000 instances can run concurrently) which processes a batch of up to 10 of said relevant data

I am happy to work off alternate proposals, but let's avoid the hand-waviness of "a design constraint".

---

Also, we should _absolutely not_ be performing a full ingestion back-fill if the goal is to produce different artefacts in the internal storage of the Construct Hub. Re-processing the discovery should only be necessary if we change the "is relevant" criteria, or if we decide to start acquiring more information that isn't present in the tarball itself.

[the remark this answers is not in the document]
Comment (eladb) [reply src/backend/discovery/index.ts:89]: Good point about not needing to reingest all of npm in order to backfill processing output. What's our design for decoupling these steps so it's trivial to upgrade the processing code and quickly update all our artifacts.

If we have a good story there I am comfortable with npm ingestion not being super scalable.

(Still feels like a bit of a tricky design if our internal hub customers would need to wait so long to start up an instance)

File: src/backend/discovery/npm-catalog-follower.lambda.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (NetaNir) [reply src/backend/discovery/npm-catalog-follower.lambda.ts:160]: fineeeeeee, but we wont remember to delete this and we will have to live with this forever! 

My take or leave it take (aka 💅 ), if the manual action is a simple as purging all data, then it is not worth it to add in the code.

File: src/backend/discovery/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (NetaNir) [reply src/backend/discovery/index.ts:89]: We talked about this offline and decided that for now (developer preview) we are ok with the limitation of 1 concurrency for the discovery function. We are going to reduce the scope of the discovery function to only scan the npmjs change endpoint. 

For the rest of the ingestion pipeline, here is what we can do (discussion required), the discovery function will post an sqs notification with the range/list of packages it had processed in a single batch, the consumer lambda will only process the packages in that range, ensuring no duplicate work between concurrent executions. This way we can slice the work in a way that allows scaling the consumer lambda.

File: src/backend/discovery/follow.lambda.ts [hunk 5]
```diff
@@ -37,13 +37,13 @@ export async function handler(event: ScheduledEvent, context: Context) {
 
   const denyList = await DenyListClient.newClient();
 
-  const initialMarker = await loadLastTransactionMarker(1_800_000 /* @aws-cdk/cdk initial release was at 1_846_709 */);
+  const marker = await loadLastTransactionMarker(1_800_000 /* @aws-cdk/cdk initial release was at 1_846_709 */);
```
Comment (RomainMuller) [inline src/backend/discovery/follow.lambda.ts:40]: We probably can change the initial marker to this one (`cdk-constants`'s latest release, which is from December 30, 2019).

```suggestion
  const marker = await loadLastTransactionMarker(4_401_405 /* around December 30, 2019 */);
```

**Comment by RomainMuller** [comment]
This looks good to me (I mean I wrote parts of this, so this isn't entirely surprising, either)

**Review by iliapolo — APPROVED** [review]
